File: lib/StateUI/Tests/StateUITests/JourneyCodecTests.swift
// SPDX-FileCopyrightText: 2026 Paweł Krzywdziński and Contributors
// SPDX-License-Identifier: Apache-2.0

import XCTest
@testable import StateUI

/// The lanes a walked value crosses in, and the batch every value crosses in,
/// held for every runtime that reads the Wire: `journey-lanes.txt` and
/// `state-batches.txt` are written from this side's own `JourneyLanes` and
/// `StateBatch`, and the MAUI host's `JourneyCodecTests.cs` reads both.
final class JourneyCodecTests: XCTestCase {
    /// Every journey in the table lies in the lanes the fixture holds.
    func testTheFixtureHoldsTheLanesThisSideWrites() throws {
        try Self.check("journey-lanes.txt", Self.journeyTable())
    }

    /// Every batch in the table lies in the bytes the fixture holds.
    func testTheFixtureHoldsTheBatchesThisSideWrites() throws {
        try Self.check("state-batches.txt", Self.batchTable())
    }

    /// A batch reads back as the writes it was made of.
    func testABatchReadsBackAsItsEntries() {
        for writes in Self.batches {
            let (read, complete) = StateBatch.decode(StateBatch.encode(writes))

            XCTAssertTrue(complete)
            XCTAssertEqual(read, writes)
        }
    }

    /// A batch cut short keeps every write before the cut, and says it was cut.
    func testABatchCutShortKeepsWhatCameBeforeTheCut() throws {
        let writes = try XCTUnwrap(Self.batches.last)
        let (read, complete) = StateBatch.decode(Array(StateBatch.encode(writes).dropLast()))

        XCTAssertFalse(complete)
        XCTAssertEqual(read, Array(writes.dropLast()))
    }

    // MARK: - The journeys

    private static func journeyTable() -> String {
        var lines = [
            "# Walked values as they lie in their lanes: where the value is, where it is",
            "# going, how fast (per second), the law's three lanes, the waiter and the",
            "# stops. Written by JourneyCodecTests.swift from JourneyLanes with",
            "# STATEUI_UPDATE_FIXTURES=1; read by JourneyCodecTests.cs.",
            "#",
            "# journey <width> value <lanes> destination <lanes> velocity <lanes>",
            "#   law <none|inherited|eased|spring|custom> <milliseconds> <curve|damping>",
            "#   waiter <id> stops <count> lanes <lanes>",
        ]

        lines.append(journey(
            0.25, to: 1.0, speed: 0.0, .eased(250, .sineInOut), waiter: -3))
        lines.append(journey(
            1.0, to: 1.0, speed: 0.0, .inherited, stops: 2))
        lines.append(journey(
            0.5, to: 0.0, speed: -2.5, .spring(response: 400, damping: 0.6), stops: 1))
        lines.append(journey(
            point(10, 20), to: point(30, -40), speed: point(5, -5), .none))
        lines.append(journey(
            rect(0, 0, 100, 50), to: rect(10, 20, 200, 80), speed: rect(0, 0, 0, 0),
            .custom, waiter: -7))
        lines.append(journey(
            rect(1, 2, 3, 4), to: rect(4, 3, 2, 1), speed: rect(-1, 1, -1, 1),
            .spring(response: 250, damping: 1), waiter: -1, stops: 5))

        return lines.joined(separator: "\n") + "\n"
    }

    /// One journey's line: its parts as this side holds them, and its lanes.
    private static func journey<Value: Walked>(
        _ value: Value,
        to destination: Value,
        speed velocity: Value,
        _ motion: Motion,
        waiter completion: Double = 0,
        stops stopped: Double = 0
    ) -> String {
        var lanes = JourneyLanes(value, motion: motion)
        lanes.destination = destination
        lanes.velocity = velocity
        lanes.completion = completion
        lanes.stopped = stopped

        return "journey \(Value.lanes)"
            + " value \(list(numbers(value)))"
            + " destination \(list(numbers(destination)))"
            + " velocity \(list(numbers(velocity)))"
            + " law \(law(motion))"
            + " waiter \(completion) stops \(stopped)"
            + " lanes \(list(numbers(lanes)))"
    }

    /// A law as the table names it, from the motion itself.
    private static func law(_ motion: Motion) -> String {
        if motion.isInherited { return "inherited 0 0.0" }
        if motion.isCustom { return "custom 0 0.0" }
        if motion.law == .eased && motion.millis == 0 { return "none 0 0.0" }

        return motion.law == .spring
            ? "spring \(motion.millis) \(motion.factor)"
            : "eased \(motion.millis) \(Double(motion.curve.rawValue))"
    }

    private static func numbers<Value: StateValue>(_ value: Value) -> [Double] {
        guard case .lanes(let lanes) = value.carried else { return [] }
        return lanes
    }

    private static func point(_ x: Double, _ y: Double) -> Point {
        Point(carried: .lanes([x, y]))!
    }

    private static func rect(_ x: Double, _ y: Double, _ width: Double, _ height: Double) -> Rect {
        Rect(carried: .lanes([x, y, width, height]))!
    }

    // MARK: - The batches

    /// Batches of every shape a crossing has: none, one, several, a number past
    /// sixteen bits, a mask past thirty-two, text, and a value with no bytes.
    private static let batches: [[StateBatch.Write]] = [
        [],
        [write(1, mask: 0b1, .lanes([0.5]))],
        [
            write(2, mask: 0b111, JourneyLanes(0.25, motion: .eased(250, .sineInOut)).carried),
            write(9, mask: 1, .text("héllo")),
        ],
        [
            write(3, mask: (1 << 40) | 1, .lanes([10, 20, 200, 80])),
            write(70_000, mask: ~0, .lanes([1, -2])),
            write(12, mask: 0, .text("")),
        ],
    ]

    private static func write(_ number: Int32, mask: UInt64, _ carried: StateCarried) -> StateBatch.Write {
        StateBatch.Write(number: number, mask: mask, bytes: StateImage.bytes(of: carried))
    }

    private static func batchTable() -> String {
        var lines = [
            "# Batches of state values as they cross a host's boundary, both ways:",
            "# [count: U16], then per write [number: I32][mask: U64][length: U32] and the",
            "# bytes, little-endian. Written by JourneyCodecTests.swift from StateBatch",
            "# with STATEUI_UPDATE_FIXTURES=1; read and written by JourneyCodecTests.cs.",
            "#",
            "# batch <bytes>",
            "# write <number> <mask> <bytes, or - for none>",
            "# cut <the last batch less its last byte> keeps <writes>",
        ]

        for writes in batches {
            lines.append("batch \(hex(StateBatch.encode(writes)))")

            for write in writes {
                let bytes = write.bytes.isEmpty ? "-" : hex(write.bytes)
                lines.append("write \(write.number) \(hex(mask: write.mask)) \(bytes)")
            }
        }

        if let last = batches.last {
            let cut = Array(StateBatch.encode(last).dropLast())
            let kept = StateBatch.decode(cut).writes.count
            lines.append("cut \(hex(cut)) keeps \(kept)")
        }

        return lines.joined(separator: "\n") + "\n"
    }

    private static func hex(_ bytes: [UInt8]) -> String {
        bytes.map { byte in
            let digits = String(byte, radix: 16)
            return byte < 16 ? "0" + digits : digits
        }.joined()
    }

    private static func hex(mask: UInt64) -> String {
        let digits = String(mask, radix: 16)
        return String(repeating: "0", count: 16 - digits.count) + digits
    }

    // MARK: - The check

    private static func list(_ values: [Double]) -> String {
        values.map { "\($0)" }.joined(separator: ",")
    }

    private static func check(_ name: String, _ written: String) throws {
        let fixture = Fixtures.directory.appendingPathComponent(name)

        if Fixtures.updating {
            try written.write(to: fixture, atomically: true, encoding: .utf8)
            return
        }

        let expected = try String(contentsOf: fixture, encoding: .utf8)

        XCTAssertEqual(
            expected, written,
            """
            \(name) no longer holds what this side writes. If the layout changed \
            on purpose, run the tests again with STATEUI_UPDATE_FIXTURES=1 and \
            read the diff of \(name).
            """)
    }
}

File: lib/StateUI.Maui/Tests/JourneyCodecTests.cs
// SPDX-FileCopyrightText: 2026 Paweł Krzywdziński and Contributors
// SPDX-License-Identifier: Apache-2.0

// The lanes a walked value crosses in and the batch every value crosses in, as
// the core writes them. JourneyCodecTests.swift writes journey-lanes.txt from
// its own JourneyLanes and state-batches.txt from its own StateBatch; this
// reads both through this side's JourneyCodec and StateBatch.

using System.Globalization;
using StateUI.Maui.Protocol;

namespace StateUI.Maui.Tests;

public class JourneyCodecTests
{
    /// <summary>
    /// Every journey the core wrote reads here as the parts it was written
    /// from: where the value is, where it is going, how fast, the law, the
    /// waiter and the stops.
    /// </summary>
    [Fact]
    public void EveryJourneyInTheCoresTableReadsAsItsParts()
    {
        int read = 0;

        foreach (string[] words in Lines("journey-lanes.txt"))
        {
            int width = int.Parse(After(words, "journey"), CultureInfo.InvariantCulture);
            double[] lanes = Numbers(After(words, "lanes"));
            string line = string.Join(' ', words);

            Assert.True(JourneyCodec.Count(width) == lanes.Length, $"{line}: {lanes.Length} lanes");
            Assert.Equal(Numbers(After(words, "value")), JourneyCodec.ValueOf(lanes, width));
            Assert.Equal(Numbers(After(words, "destination")), JourneyCodec.DestinationOf(lanes, width));
            Assert.Equal(Numbers(After(words, "velocity")), JourneyCodec.VelocityOf(lanes, width));
            Assert.Equal(Number(After(words, "waiter")), lanes[JourneyCodec.Waiter(width)]);
            Assert.Equal(Number(After(words, "stops")), lanes[JourneyCodec.Stops(width)]);

            int law = Array.IndexOf(words, "law");
            (JourneyCodec.Law kind, HostMotion motion) = JourneyCodec.MotionAt(lanes, JourneyCodec.LawAt(width));

            Assert.True(
                string.Equals(words[law + 1], kind.ToString(), StringComparison.OrdinalIgnoreCase),
                $"{line}: read as {kind}");

            switch (kind)
            {
                case JourneyCodec.Law.Eased:
                    Assert.Equal(HostMotion.Law.Eased, motion.Kind);
                    Assert.Equal(uint.Parse(words[law + 2], CultureInfo.InvariantCulture), motion.Millis);
                    Assert.Equal((HostEasing)(int)Number(words[law + 3]), motion.Curve);
                    break;

                case JourneyCodec.Law.Spring:
                    Assert.Equal(HostMotion.Law.Spring, motion.Kind);
                    Assert.Equal(uint.Parse(words[law + 2], CultureInfo.InvariantCulture), motion.Millis);
                    Assert.Equal(Number(words[law + 3]), motion.Factor);
                    break;

                default:
                    Assert.True(motion.Instant, $"{line}: {kind} moves");
                    break;
            }

            read++;
        }

        Assert.True(read >= 5, "the table holds almost no journeys");
    }

    /// <summary>
    /// Every batch the core wrote reads here as its writes, and those writes
    /// lie in the core's own bytes again; a batch cut short keeps the writes
    /// before the cut, as the core's reader does.
    /// </summary>
    [Fact]
    public void EveryBatchInTheCoresTableReadsAndWritesAsTheCoresDoes()
    {
        List<(byte[] Bytes, List<(int Number, ulong Mask, byte[] Bytes)> Writes)> batches = [];
        byte[]? cut = null;
        int keeps = 0;

        foreach (string[] words in Lines("state-batches.txt"))
        {
            switch (words[0])
            {
                case "batch":
                    batches.Add((Convert.FromHexString(words[1]), []));
                    break;

                case "write":
                    batches[^1].Writes.Add((
                        int.Parse(words[1], CultureInfo.InvariantCulture),
                        ulong.Parse(words[2], NumberStyles.HexNumber, CultureInfo.InvariantCulture),
                        words[3] == "-" ? [] : Convert.FromHexString(words[3])));
                    break;

                case "cut":
                    cut = Convert.FromHexString(words[1]);
                    keeps = int.Parse(words[3], CultureInfo.InvariantCulture);
                    break;
            }
        }

        Assert.True(batches.Count >= 4, "the table holds almost no batches");

        foreach ((byte[] bytes, List<(int Number, ulong Mask, byte[] Bytes)> writes) in batches)
        {
            Same(writes, StateBatch.Read(bytes));
            Assert.Equal(bytes, StateBatch.Bytes(writes));
        }

        Assert.NotNull(cut);
        Same(batches[^1].Writes[..keeps], StateBatch.Read(cut));
    }

    private static void Same(
        List<(int Number, ulong Mask, byte[] Bytes)> expected,
        List<(int Number, ulong Mask, byte[] Bytes)> read)
    {
        Assert.Equal(expected.Count, read.Count);

        for (int index = 0; index < expected.Count; index++)
        {
            Assert.Equal(expected[index].Number, read[index].Number);
            Assert.Equal(expected[index].Mask, read[index].Mask);
            Assert.Equal(expected[index].Bytes, read[index].Bytes);
        }
    }

    /// <summary>Every line of a table that is not a comment, as its words.</summary>
    private static IEnumerable<string[]> Lines(string name) =>
        File.ReadAllLines(Path.Combine(Fixtures.Directory, name))
            .Where(line => line.Length > 0 && !line.StartsWith('#'))
            .Select(line => line.Split(' '));

    /// <summary>The word after <paramref name="key"/>.</summary>
    private static string After(string[] words, string key) => words[Array.IndexOf(words, key) + 1];

    private static double Number(string word) =>
        double.Parse(word, NumberStyles.Float, CultureInfo.InvariantCulture);

    private static double[] Numbers(string word) => [.. word.Split(',').Select(Number)];
}
